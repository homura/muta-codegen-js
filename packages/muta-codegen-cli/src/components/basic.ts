import { codegen as graphqlCodegen } from '@graphql-codegen/core';
import { CodegenPlugin, Types } from '@graphql-codegen/plugin-helpers/types';
import { buildSchema, parse, printSchema } from 'graphql';

interface Options {
  schemaSource: string;
  pluginsMap: {
    [name: string]: CodegenPlugin;
  };
  plugins: Types.ConfiguredPlugin[];
}

export function codegen(options: Options): Promise<string> {
  if (!options.schemaSource) return Promise.resolve('');

  return graphqlCodegen({
    schema: parse(printSchema(buildSchema(options.schemaSource))),
    plugins: options.plugins,
    pluginMap: options.pluginsMap,
    filename: '',
    documents: [],
    config: {
      scalars: {
        Address: 'Address',
        u64: 'u64',
        Hash: 'Hash',
        Uint64: 'Uint64',
        Null: "''",
        Hex: 'string',
        Uint32: 'string',
      },
      skipTypename: true,
    },
  });
}
