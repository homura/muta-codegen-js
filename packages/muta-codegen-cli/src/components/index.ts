export { generateEventCode } from './event';
export { generateServiceBindingCode } from './service';
export { generatePackageJSONCode } from './packageJSON';
