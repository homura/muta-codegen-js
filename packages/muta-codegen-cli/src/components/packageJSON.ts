interface PackageJSONOptions {
  name: string;
  mutaSDKVersion: string;
}

export function generatePackageJSONCode(
  options: Partial<PackageJSONOptions>,
): string {
  return `{
  "name": "${options.name}",
  "main": "lib/index.js",
  "scripts": {
    "build": "tsc --outDir lib src/**/*.ts"
  },
  "dependencies": {
    "@mutajs/types": "${options.mutaSDKVersion}",
    "@mutajs/service": "${options.mutaSDKVersion}"
  },
  "devDependencies": {
    "typescript": "^3.8.0"
  }
} `;
}
