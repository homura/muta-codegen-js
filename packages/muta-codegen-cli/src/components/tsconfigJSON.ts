export function generateTsconfigJSONCode(): string {
  return `{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es5",
    "declaration": true,
    "outDir": "lib"
  },
  "exclude": [
    "node_modules"
  ],
  "include": ["./src"]
}`;
}
