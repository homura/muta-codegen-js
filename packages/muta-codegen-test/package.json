{
  "name": "muta-codegen-test",
  "version": "0.0.1",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "private": true,
  "files": [
    "lib"
  ],
  "scripts": {
    "clean": "rimraf lib",
    "test": "jest",
    "test:watch": "jest --watch",
    "coverage": "jest --coverage",
    "lint": "eslint",
    "lint:fix": "eslint --fix",
    "precommit": "lint-staged"
  },
  "dependencies": {
  }
}
