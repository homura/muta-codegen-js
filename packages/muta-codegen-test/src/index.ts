export { AssetServiceSchema } from './mock/AssetService.graphql';
export { AssetEventSchema } from './mock/AssetEvent.graphql';
