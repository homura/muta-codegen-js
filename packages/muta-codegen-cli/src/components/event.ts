import * as typescriptPlugin from '@graphql-codegen/typescript';
import { event as eventPlugin } from 'graphql-codegen-plugin-mutasdk';
import { codegen } from './basic';

export async function generateEventCode(eventSource: string): Promise<string> {
  return codegen({
    pluginsMap: {
      typescript: typescriptPlugin,
      event: eventPlugin,
    },
    plugins: [{ event: {} }, { typescript: {} }],
    schemaSource: eventSource,
  });
}
