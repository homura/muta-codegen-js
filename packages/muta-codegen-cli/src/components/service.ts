import * as typescriptPlugin from '@graphql-codegen/typescript';
import { service as servicePlugin } from 'graphql-codegen-plugin-mutasdk';
import { codegen } from './basic';

export async function generateServiceBindingCode(
  serviceName: string,
  schemaSource: string,
): Promise<string> {
  return codegen({
    pluginsMap: {
      typescript: typescriptPlugin,
      service: servicePlugin,
    },
    plugins: [{ service: { service: serviceName } }, { typescript: {} }],
    schemaSource,
  });
}
