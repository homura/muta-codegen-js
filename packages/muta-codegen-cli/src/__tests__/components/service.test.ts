import { AssetEventSchema, AssetServiceSchema } from 'muta-codegen-test';
import { generateServiceBindingCode } from '../../components';

test('asset service', async () => {
  const code = await generateServiceBindingCode('asset', AssetServiceSchema);
  expect(code).toMatchSnapshot();
});

test('metadata service with empty args get_metadata: Metadata', async () => {
  expect(
    await generateServiceBindingCode('metadata', AssetEventSchema),
  ).toMatchSnapshot();
});
