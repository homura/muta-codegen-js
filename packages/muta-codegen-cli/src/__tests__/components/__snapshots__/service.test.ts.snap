// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asset service 1`] = `
"// generated by muta-codegen, please do not edit manually
/* eslint-disable */
import { createBindingClass, query, mutation } from \\"@mutajs/service\\";
import { u64, Hash, Address, Uint64 , Bytes, u32, Vec, Null } from \\"@mutajs/types\\";

export type Maybe<T> = T | null;

export const AssetService = createBindingClass(
  'asset',
  {
    get_asset: query<GetAssetPayload, Asset>(),
    get_balance: query<GetBalancePayload, GetBalanceResponse>(),
    get_allowance: query<GetAllowancePayload, GetAllowanceResponse>(),
  },
  {
    create_asset: mutation<CreateAssetPayload, Asset>(),
    transfer: mutation<TransferPayload, null>(),
    approve: mutation<ApprovePayload, null>(),
    transfer_from: mutation<TransferFromPayload, null>(),
  }
);

/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Address: Address;
  Hash: Hash;
  Uint64: Uint64;
  Null: '';
};

export type Mutation = {
  create_asset: Asset;
  transfer?: Maybe<Scalars['Null']>;
  approve?: Maybe<Scalars['Null']>;
  transfer_from?: Maybe<Scalars['Null']>;
};


export type MutationCreate_AssetArgs = {
  payload: CreateAssetPayload;
};


export type MutationTransferArgs = {
  payload: TransferPayload;
};


export type MutationApproveArgs = {
  payload: ApprovePayload;
};


export type MutationTransfer_FromArgs = {
  payload: TransferFromPayload;
};

export type Query = {
  get_asset: Asset;
  get_balance: GetBalanceResponse;
  get_allowance: GetAllowanceResponse;
};


export type QueryGet_AssetArgs = {
  payload: GetAssetPayload;
};


export type QueryGet_BalanceArgs = {
  payload: GetBalancePayload;
};


export type QueryGet_AllowanceArgs = {
  payload: GetAllowancePayload;
};


export type ApprovePayload = {
  asset_id: Scalars['Hash'];
  to: Scalars['Address'];
  value: Scalars['Uint64'];
};

export type Asset = {
  id: Scalars['Hash'];
  name: Scalars['String'];
  symbol: Scalars['String'];
  supply: Scalars['Uint64'];
  issuer: Scalars['Address'];
};

export type CreateAssetPayload = {
  name: Scalars['String'];
  symbol: Scalars['String'];
  supply: Scalars['Uint64'];
};

export type GetAllowancePayload = {
  asset_id: Scalars['Hash'];
  grantor: Scalars['Address'];
  grantee: Scalars['Address'];
};

export type GetAllowanceResponse = {
  asset_id: Scalars['Hash'];
  grantor: Scalars['Address'];
  grantee: Scalars['Address'];
  value: Scalars['Uint64'];
};

export type GetAssetPayload = {
  id: Scalars['Hash'];
};

export type GetBalancePayload = {
  asset_id: Scalars['Hash'];
  user: Scalars['Address'];
};

export type GetBalanceResponse = {
  asset_id: Scalars['Hash'];
  user: Scalars['Address'];
  balance: Scalars['Uint64'];
};


export type TransferFromPayload = {
  asset_id: Scalars['Hash'];
  sender: Scalars['Address'];
  recipient: Scalars['Address'];
  value: Scalars['Uint64'];
};

export type TransferPayload = {
  asset_id: Scalars['Hash'];
  to: Scalars['Address'];
  value: Scalars['Uint64'];
};


"
`;

exports[`metadata service with empty args get_metadata: Metadata 1`] = `
"// generated by muta-codegen, please do not edit manually
/* eslint-disable */
import { createBindingClass, query, mutation } from \\"@mutajs/service\\";
import { u64, Hash, Address, Uint64 , Bytes, u32, Vec, Null } from \\"@mutajs/types\\";

export type Maybe<T> = T | null;

export const MetadataService = createBindingClass(
  'metadata',
  {
    
  },
  {
    
  }
);

/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Address: Address;
  Hash: Hash;
  Uint64: Uint64;
};


export type ApproveEvent = {
  asset_id: Scalars['Hash'];
  grantor: Scalars['Address'];
  grantee: Scalars['Address'];
  value: Scalars['Uint64'];
};

export type Asset = {
  id: Scalars['Hash'];
  name: Scalars['String'];
  symbol: Scalars['String'];
  supply: Scalars['Uint64'];
  issuer: Scalars['Address'];
};


export type TransferEvent = {
  asset_id: Scalars['Hash'];
  from: Scalars['Address'];
  to: Scalars['Address'];
  value: Scalars['Uint64'];
};

export type TransferFromEvent = {
  asset_id: Scalars['Hash'];
  caller: Scalars['Address'];
  sender: Scalars['Address'];
  recipient: Scalars['Address'];
  value: Scalars['Uint64'];
};


export type Event = Asset | TransferEvent | ApproveEvent | TransferFromEvent;
"
`;
